### PR TITLE
typeahead: custom remote entity typeahead component

### PR DESCRIPTION
**Pull request by zannkukai**
Creates a remote typeahead component that allows to split suggestions
into several column. Suggestions box includes a button to create a new
local entity if current logged user has permissions to create it.

Co-Authored-By: Renaud Michotte <renaud.michotte@gmail.com>

## requirements

https://github.com/rero/ng-core/pull/558

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: projects/admin/src/app/classes/typeahead/mef-typeahead.ts [hunk 1]
```diff
@@ -81,20 +109,72 @@ export class MefTypeahead implements ITypeahead {
...
+    return output;
+  }
+
+  /**
+   * Manage entity resolution error.
+   * @param error: the triggered error
+   * @param value: the entity value causing the resolution error.
+   */
+  private _handleEntityResolutionError(error: Error, value: string): Observable<string> {
+    console.error(`getValueAsHTML :: Unable to resolve entity :: ${value}`, error);
```
Comment (Garfield-fr) [inline projects/admin/src/app/classes/typeahead/mef-typeahead.ts:172]: Remove ?

File: projects/admin/src/app/classes/typeahead/mef-typeahead.ts [hunk 2]
```diff
@@ -111,36 +191,75 @@ export class MefTypeahead implements ITypeahead {
...
+    });
+    return suggestions;
   }
 
+  /**
+   * Build suggestions for local entities API call.
+   * @param hits: the hits returned by the API call.
+   * @returns Suggestions corresponding to hits.
+   */
+  private _buildLocalSuggestions(hits: any): Array<SuggestionMetadata> {
```
Comment (Garfield-fr) [inline projects/admin/src/app/classes/typeahead/mef-typeahead.ts:248]: `SuggestionMetadata[]`

File: projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  templateUrl: './entity-typeahead.component.html',
+  styleUrls: ['./entity-typeahead.component.scss']
+})
+export class EntityTypeaheadComponent extends FieldType implements OnInit {
+
+  // COMPONENT ATTRIBUTES =====================================================
+  /** searched term */
+  searchTerm: string = null;
+  /** entity type filter options */
+  entityTypeFilters: Array<EntityTypeFilter> = [];
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts:21]: `EntityTypeFilter[]`

File: projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  /** is the suggestions are loading ? */
+  isSuggestionsLoading: boolean = false;
+
+  /** observable on entity suggestions */
+  suggestion$: Observable<Array<Array<SuggestionMetadata | string>>>;
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts:26]: The same as above

File: projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  /** observable on representation of the formControl value. */
+  valueAsHTML$: Observable<string>;
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts:28]: The same as above

File: projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  /** suggestions data from remote typeahead service */
+  suggestionSections: Array<Array<TypeaheadMatch>> = [];
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts:30]: The same as above

File: projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  /** Build the entity type filters based on `templateOptions` field section. */
+  private _buildEntityTypeFilters(): void {
+    this.to?.filters?.options.map(option => this.entityTypeFilters.push({
+      ...option,
+      ...{selected: option.value === this.to?.filters?.default}
+    }));
+  }
+
+  /** Split suggestion from typeahead service by column. */
+  private _splitSuggestionsByColumn(suggestions: Array<SuggestionMetadata>): Array<Array<SuggestionMetadata>> {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts:119]: The same as above

File: projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        sections[columnIdx] = [];
+      }
+      delete suggestion?.column;
+      sections[columnIdx].push(suggestion);
+    });
+    return sections.filter(section => section);  // remove empty slot/column from array
+  }
+
+  /** Order each suggestion for each section. */
+  private _orderSuggestions(suggestionSections: Array<Array<SuggestionMetadata>>): Array<Array<SuggestionMetadata>> {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts:136]: The same as above

File: projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    return suggestionSections.map(section =>
+      section.sort((a, b) =>
+        (a.group === b.group)
+          ? a.label.localeCompare(b.label)
+          : a.group.localeCompare(b.group))
+    );
+  }
+
+  /** Create group header for each suggestion sections. */
+  private _createSuggestionGroupHeader(suggestionSections: Array<Array<SuggestionMetadata>>): Array<Array<TypeaheadMatch>> {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/formly/type/entity-typeahead/entity-typeahead.component.ts:146]: The same as above

**Review by Garfield-fr — APPROVED** [review]

**Comment by lauren-d** [comment]
you have a type in your commit message:
`typeahead: custom remote entity typeahad component`
=> typeahead: custom remote entity typeah**e**ad component

File: projects/admin/src/app/classes/typeahead/mef-typeahead.ts [hunk 10]
```diff
@@ -111,36 +191,75 @@ export class MefTypeahead implements ITypeahead {
...
-            throw e;
+          case 400: return of([]);
+          default: throw e;
         }
       })
     );
+    // Return the concatenation of both API request.
+    return forkJoin([remoteSuggestions$, localSuggestions$]).pipe(
+      map(([remoteHits, localHits]) => [...remoteHits, ...localHits])
+    );
```
Comment (lauren-d) [inline projects/admin/src/app/classes/typeahead/mef-typeahead.ts:223]: general remark: why not to have one api like `api/entities/suggestions` to retrieve all formatted suggestions ?
😜

**Review by lauren-d — APPROVED** [review]

File: projects/admin/src/app/classes/typeahead/mef-typeahead.ts (hunk 1 above)
Comment (zannkukai) [reply projects/admin/src/app/classes/typeahead/mef-typeahead.ts:172]: No. It sounds good (for me ?) to see an error in the console in this case.

**Pull request merged**